src/timesync/timesyncd-server.c

```diff
@@ -175,3 +178,39 @@ void server_name_flush_addresses(ServerName *n) {
         while (n->addresses)
                 server_address_free(n->addresses);
 }
+
+int server_name_parse_port(ServerName *n) {
+        const char *sqr = "]", *col = ":";
+        char *ret_sqr, *ret_col, *last_col;
+
+        ret_sqr = strrchr(n->string, *sqr);
+        ret_col = strchr(n->string, *col);
+        last_col = strrchr(n->string, *col);
+
+        if (ret_sqr == NULL && ret_col == NULL) { // server.domain or I.P.v.4
+                return 0;
+        } else if (ret_sqr == NULL && strlen(ret_col) == strlen(last_col)) { // has no ']' and exactly one ":"
+                const char *word = strdupa_safe(n->string);
+                (void) extract_first_word(&word, &n->string, col, 0);
+                (void) extract_first_word(&word, &n->overridden_port, col, 0);
+                log_debug("Matched single port: %s / %s", n->string, n->overridden_port);
+                return 1;
+        } else if (ret_sqr == NULL && ret_col != NULL && last_col != NULL && strlen(ret_col) != strlen(last_col)) {
+                // naked IP::v:6, no ']' and more than one ':'
+                return 0;
+        } else if (ret_sqr != NULL && strlen(ret_sqr) == strlen(last_col)+1) { // [IP::v:6]:port with "]:" substring
+                const char *word = strdupa_safe(n->string);
+                (void) extract_first_word(&word, &ret_col, sqr, 0);
+                n->string = strncat(ret_col, sqr, strlen(sqr)+1);
+                word = last_col;
+                (void) extract_first_word(&word, &n->overridden_port, col, 0);
+                log_debug("Matched [IP::v:6]:port, output  %s / %s", n->string, n->overridden_port);
+                return 2;
+        } else if (ret_sqr != NULL && last_col != NULL && strlen(ret_sqr) != strlen(last_col)+1) {
+                // [IP::v:6] without port -- no "]:" substring
+                return 0;
+        } else {
+                assert_not_reached();
+                return -1;
```

Comment:
do not make up errors, we return negative Exyz errors, errno style.

Reply:
also, assert that some piece of code is not reached followed by code doesn't really make much sense.